### DISCUSSION TITLE
Working in debug mode on LibIGL project

### DISCUSSION THREAD

**Discussion by drathin4** [libigl/libigl#1801] [Q&A]
I have created a LibIGL project and I compiled the VisualStudio code in both debug and release versions. The release version is working correctly, but I'm having issues with the debug version. 

When I double click the example.exe file in project\build\Debug folder, it says, 
"Debug Error! abort() has been called!". 

Also when I run F5 in VisualStudio it assserts,
"Unable to start program project\build\x64\ALL_BUILD. The system cannot find the file specified.".

The program runs correctly in release version. Later in the project, I will have to do a lot of debugging, hence need the debug version to work correctly.

**Comment by jdumas** [comment]
We need more information to answer your question. E.g. where exactly does the error occur in Debug mode? What is the stacktrace? What executable are you running? Is it the libigl-example-project, a libigl tutorial, your own project? An error in Debug mode but not in Release is usually indicative of an `assert()` check that was violated, so you need to check what is going on.

**Reply by drathin4** [reply]
This is my own LibIGL project. I'm working on MeshBoolean project.

The moment I hit F5 in VisualStudio, I immediately get the error.

![Capture](https://user-images.githubusercontent.com/79855313/117018859-85c18180-ad12-11eb-917f-75e9402aecac.PNG)

I can see the following build trace in VisualStudio.

6>example.vcxproj -> C:\Rathin\Projects\MeshBoolean\build\Debug\example.exe
6>Done building project "example.vcxproj".
7>------ Skipped Build: Project: ALL_BUILD, Configuration: Debug x64 ------
7>Project not selected to build for this solution configuration 
========== Build: 1 succeeded, 0 failed, 3 up-to-date, 6 skipped ==========

Also, after commenting out the problematic line in the code, I was able to run the run example.exe from project\build\Debug folder. The problematic line of code is the readSTL() call. I need to understand, why it's not able to read the STL file and hence need to debug into readSTL.cpp file in VisualStudio (which unfortunately I can't because of the error I showed above).

**Comment by jdumas** [comment]
For Visual Studio you need to set the startup project to the executable you want to run. `ALL_BUILD` is just a meta-project that will build your entire Visual Studio solution.

**Reply by drathin4** [reply]
I managed to change the startup project and now it is working. 

But I want to debug into readSTL.cpp file, hence I added that as a source file to VisualStudio solution. On building the solution, I'm getting errors in readSTL.cpp file which I have not altered. 

6>C:\Rathin\Projects\MeshBoolean\libigl\include\igl\readSTL.cpp(20,17): error C2995: 'bool igl::readSTL(std::istream &,Eigen::PlainObjectBase<Derived> &,Eigen::PlainObjectBase<DerivedF> &,Eigen::PlainObjectBase<DerivedN> &)': function template has already been defined
6>C:\Rathin\Projects\MeshBoolean\libigl\include\igl\readSTL.h(42): message : see declaration of 'igl::readSTL'
6>C:\Rathin\Projects\MeshBoolean\libigl\include\igl\readSTL.cpp(45,17): error C2084: function 'bool igl::is_stl_binary(std::istream &)' already has a body
6>C:\Rathin\Projects\MeshBoolean\libigl\include\igl\readSTL.cpp(45): message : see previous definition of 'is_stl_binary'
6>C:\Rathin\Projects\MeshBoolean\libigl\include\igl\readSTL.cpp(77,17): error C2995: 'bool igl::read_stl_ascii(std::istream &,std::vector<std::array<TypeV,3>,std::allocator<std::array<TypeV,3>>> &,std::vector<std::array<TypeF,3>,std::allocator<std::array<TypeF,3>>> &,std::vector<std::array<TypeN,3>,std::allocator<std::array<TypeN,3>>> &)': function template has already been defined
6>C:\Rathin\Projects\MeshBoolean\libigl\include\igl\readSTL.cpp(77): message : see declaration of 'igl::read_stl_ascii'
6>C:\Rathin\Projects\MeshBoolean\libigl\include\igl\readSTL.cpp(192,17): error C2995: 'bool igl::read_stl_binary(std::istream &,std::vector<std::array<TypeV,3>,std::allocator<std::array<TypeV,3>>> &,std::vector<std::array<TypeF,3>,std::allocator<std::array<TypeF,3>>> &,std::vector<std::array<TypeN,3>,std::allocator<std::array<TypeN,3>>> &)': function template has already been defined
6>C:\Rathin\Projects\MeshBoolean\libigl\include\igl\readSTL.cpp(192): message : see declaration of 'igl::read_stl_binary'
6>C:\Rathin\Projects\MeshBoolean\libigl\include\igl\readSTL.cpp(285,17): error C2995: 'bool igl::readSTL(std::istream &,std::vector<std::array<TypeV,3>,std::allocator<std::array<TypeV,3>>> &,std::vector<std::array<TypeF,3>,std::allocator<std::array<TypeF,3>>> &,std::vector<std::array<TypeN,3>,std::allocator<std::array<TypeN,3>>> &)': function template has already been defined
6>C:\Rathin\Projects\MeshBoolean\libigl\include\igl\readSTL.h(49): message : see declaration of 'igl::readSTL'
6>C:\Rathin\Projects\MeshBoolean\libigl\include\igl\readSTL.cpp(299,17): error C2995: 'bool igl::readSTL(FILE *,Eigen::PlainObjectBase<Derived> &,Eigen::PlainObjectBase<DerivedF> &,Eigen::PlainObjectBase<DerivedN> &)': function template has already been defined
6>C:\Rathin\Projects\MeshBoolean\libigl\include\igl\readSTL.h(56): message : see declaration of 'igl::readSTL'

How do I resolve this error ?

**Reply by jdumas** [reply]
You should not modify the Visual Studio solution generated by CMake. If you have an error when running your program you should use Visual Studio to step into your program, look at the stack trace and the state of your program, etc. Upon debugging Visual Studio will directly open the line that crashed your program.

**Reply by drathin4** [reply]
I removed the readSTL.cpp file from solution source files and now I'm able to debug into readSTL.cpp file. I saw the line where the code is raising an assertion. It is at `assert(input.good());`. But I don't understand how come it passed the earlier checks and then raising assertion at line 224 in the readSTL.cpp file. Following is the code snippet where the problem is happening (2nd breakpoint). It has passed the `input.good()` checks at lines 207 and 213, but raising assertion at line 224.

![image](https://user-images.githubusercontent.com/79855313/117047827-f034ea80-ad2f-11eb-8d7a-b471fd20c700.png)

**Comment by jdumas** [comment]
Well looks like the code reached the end of your STL file and was not able to read the content properly. It could be an issue with your file, or it could be an issue with our STL parser. If you can open your mesh with another program (e.g. MeshLab), then please share your STL with us so we can investigate the issue.

**Reply by drathin4** [reply]
I tried opening this STL file with an online tool [https://www.viewstl.com/](url) and was able to view it clearly. Also this file can be viewed in LibIGL using the Viewer member function `load_mesh_from_file()`. Just that it's having issues with readSTL.cpp (specifically `read_stl_binary()` function). So I think chances are that it's a bug. I'm attaching the 5 MB zipped STL file in this reply.

[0.zip](https://github.com/libigl/libigl/files/6425475/0.zip)
